A45-retry-stats.md

```diff
@@ -0,0 +1,126 @@
+| grpc.io/client/retries\_per\_call              | 1    | Number of retry or hedging attempts excluding transparent retries made during the client call. The original attempt is not counted as a retry/hedging attempt. |
+| grpc.io/client/transparent\_retries\_per\_call | 1    | Number of transparent retries made during the client call.                                                                                                     |
+| grpc.io/client/retry\_delay\_per\_call         | ms   | Total time of delay while there is no active attempt during the client call.           
+
+#### Views
+
+| View name                                      | Measure                         | Aggregation  | tags                 |
+| ---------------------------------------------- | ------------------------------- | ------------ | -------------------- |
+| grpc.io/client/retries\_per\_call              | retries\_per\_call              | distribution | grpc\_client\_method |
+| grpc.io/client/retries                         | retries\_per\_call              | sum          | grpc\_client\_method |
+| grpc.io/client/transparent\_retries\_per\_call | transparent\_retries\_per\_call | distribution | grpc\_client\_method |
+| grpc.io/client/transparent\_retries            | transparent\_retries\_per\_call | sum          | grpc\_client\_method |
+| grpc.io/client/retry\_delay\_per\_call         | retry\_delay\_per\_call         | distribution | grpc\_client\_method |
+
+
+### Tracing
+
+If retry is not enabled on the client channel, the gRPC tracing module should behave the same as it does today. If retry is enabled, the gRPC tracing module should create a tracing span for each call from the perspective of the application, and create a child span for each individual call attempt, including transparent retry attempts. The parent span is named in the form of `Sent.<service name>.<method name>`, which is the same as the case when retry is not enabled. The span for each individual call attempt is named in the form of `Attempt.<service name>.<method name>`, and is with the following additional span attributes:
+
+      `previous-rpc-attempts` : number of preceding attempts,
+      `transparent-retry`     : whether the attempt is a transparent retry.
+
+The child span should record the message events of the attempt as the call span would do in the case when retry is not enabled.
+
+## Rationale
+
+Transparent retry should also be treated as a distinct RPC with regards to the per-attempt metrics, because for OpenCensus tracing, it's natural to create a child span for transparent retry as with ordinary retry attempt, so for OpenCensus stats, we should be consistent. Also, there is no effective way to distinguish the two types of transparent retry, and one of them actually sends data to the wire, so the outbound bytes should be recorded.
```

Comment:
In what metrics are they handled the same way? They aren't included in the retries metric and the transparent_retries metric doesn't include the normal retries.

Reply:
Not sure if understand your question correctly. They handle the per-attempt metrics in the same way.

Reply:
It'd probably be more clear if the text was word-wrapped and I didn't have to comment on a full paragraph.

You talk about "treated as distinct RPC" for per-attempt metrics, but I don't actually see anything that is being kept "consistent." If we were consistent, given the description above, we would have a single counter for transparent and non-transparent retries. Instead, we treat transparent retries specially and different. I don't see any metric where non-transparent retries and transparent retries are handled the same way.

Reply:
Discussed offline. Clarified a bit and removed misleading commas.